students/grant_dowell/Lesson_04/mailroom_part2.py

```diff
@@ -0,0 +1,121 @@
+# -*- coding: utf-8 -*-
+"""
+Mailroom Part 1
+Grant Dowell
+"""
+
+# Data structure requirements:
+#  - Name, history of donation amounts
+
+# Program Utilites
+#  1) Send Thank You
+#     - Prompt for full name
+#     - if
+#  2) Create a Report
+#  3) Quit
+
+def sortby_amt(entry):
+    return sum(entry[1])
+
+def generate_letter(name=None):
+    """ Builds a letter using the global database"""
+
+    template = "Dear {name},\n\n" + \
+               "Thank you for your most recent donation of " + \
+               "${last_donation}. We greatly appreciate it.\n\n" + \
+               " ~ The Treasurer"
+    donor_info = {'name':name, 'last_donation':db[name][-1]}
+    letter = template.format(**donor_info)
+    return letter
+
+def thanks():
+    """ Generates a Thank You message for a new donation, and adds donation
+        to database """
+
+    thank_you = ""
+    vld_name = False
+    while vld_name is False:
+        name = input("Please enter a Full Name or 'List': ")
+        if name.lower() == 'list':
+            for name in db.keys():
+                print(name)
+        elif name.lower() == 'quit':
+            break
+        else:
+            vld_name = True
+            for _ in db:    # Check for name in db
+                if name.lower == name.lower:
+                    break
```

Comment:
this is incorrect - because you're using a dictionary you should utilize the efficient key look up.

I also don't understand the purpose of this loop or why you would compare two identical names which would always be true and this would actually compare function objects. This also causes the program to crash when adding new donors.
```
if db.get(name.lower())
```

Reply:
Thanks for the feedback. I also altered the db structure to use all lowercase keys and just make the input lower case from the get-go.

Reply:
👍